ecs-cli/modules/cli/servicediscovery/servicediscovery_app.go

```diff
@@ -30,138 +31,160 @@ const (
+	cfnTemplateOutputSDSARN             = "ServiceDiscoveryServiceARN"
 )
 
-const (
-	parameterKeySDSDescription                          = "SDSDescription"
-	parameterKeySDSName                                 = "SDSName"
-	parameterKeyNamespaceID                             = "NamespaceID"
-	parameterKeyDNSType                                 = "DNSType"
-	parameterKeyDNSTTL                                  = "DNSTTL"
-	parameterKeyHealthCheckCustomConfigFailureThreshold = "FailureThreshold"
-)
+// Create creates a DNS namespace (or uses an existing one) and creates a Service Discovery Service
+// The Service Discovery ARN is returned so that it can be used to enable ECS Service Discovery
+func Create(networkMode, serviceName string, c *context.ECSContext) (*string, error) {
+	cfnClient := cloudformation.NewCloudformationClient(c.CommandConfig)
 
-const (
-	CFNTemplateOutputPrivateNamespaceID = "PrivateDNSNamespaceID"
-	CFNTemplateOutputSDSARN             = "ServiceDiscoveryServiceARN"
-)
-
-const (
-	dnsRecordTypeA   = "A"
-	dnsRecordTypeSRV = "SRV"
-)
+	var ecsParamsSD *utils.ServiceDiscovery
+	if c.ECSParams != nil {
+		ecsParamsSD = &c.ECSParams.RunParams.ServiceDiscovery
+	} else {
+		ecsParamsSD = &utils.ServiceDiscovery{}
+	}
 
-var requiredParamsSDS = []string{parameterKeyNamespaceID, parameterKeySDSName, parameterKeyDNSType}
-var requiredParamsNamespace = []string{parameterKeyVPCID, parameterKeyNamespaceName}
+	return create(c.CLIContext, networkMode, serviceName, cfnClient, ecsParamsSD, c.CommandConfig)
+}
 
-// Create creates resources for service discovery and returns the ID of the Service Discovery Service
-func Create(c *cli.Context, networkMode, serviceName, clusterName string) (*string, error) {
-	rdwr, err := config.NewReadWriter()
+func create(c *cli.Context, networkMode, serviceName string, cfnClient cloudformation.CloudformationClient, ecsParamsSD *utils.ServiceDiscovery, config *config.CommandConfig) (*string, error) {
+	err := validateNotNameAndID(c, ecsParamsSD)
 	if err != nil {
 		return nil, err
 	}
-
-	commandConfig, err := config.NewCommandConfig(c, rdwr)
+	input, err := mergeSDFlagsAndInput(c, ecsParamsSD)
 	if err != nil {
 		return nil, err
 	}
-
-	cfnClient := cloudformation.NewCloudformationClient(commandConfig)
-
-	return create(c, networkMode, serviceName, clusterName, cfnClient)
-}
-
-func create(c *cli.Context, networkMode, serviceName, clusterName string, cfnClient cloudformation.CloudformationClient) (*string, error) {
-	// create namespace
-	namespaceParams := namespaceCFNParams(c)
-	if err := namespaceParams.Validate(); err != nil {
-		return nil, err
-	}
-
-	namespaceStackName := cfnStackName(privateDNSNamespaceStackNameFormat, clusterName, serviceName)
-	if _, err := cfnClient.CreateStack(cloudformation.GetPrivateNamespaceTemplate(), namespaceStackName, false, namespaceParams); err != nil {
+	err = validateMergedSDInputFields(input, networkMode)
```

Comment:
Hm, it seems a little clunky to have to run the validation twice, before and after merge, but it's fine for now.

Reply:
Yeah its annoying, but since input can from two sources (Params and flags) I need to run some validations on them individually, and then some validations on the result of merging them.